[AArch64][SME] Save VG for unwind info when changing streaming-mode

kmclaughlin-arm opened the pull request:
If a function requires any streaming-mode change, the vector granule
value must be stored to the stack and unwind info must also describe the
save of VG to this location.

This patch adds VG to the list of callee-saved registers and increases the
callee-saved stack size if the function requires streaming-mode changes.
A new type is added to RegPairInfo, which is also used to skip restoring
the register used to spill the VG value in the epilogue.

See https://github.com/ARM-software/abi-aa/blob/main/aadwarf64/aadwarf64.rst

kmclaughlin-arm committed:
[AArch64][SME] Save VG for unwind info when changing streaming-mode

If a function requires any streaming-mode change, the vector granule
value must be stored to the stack and unwind info must also describe the
save of VG to this location.

This patch adds VG to the list of callee-saved registers and increases the
callee-saved stack size in determineCalleeSaves if the function requires
streaming-mode changes. A new type is added to RegPairInfo for VG, which is
also used to skip restoring the register in the restore block.

See https://github.com/ARM-software/abi-aa/blob/main/aadwarf64/aadwarf64.rst

efriedma-quic commented:
It seems weird to be generating actual instructions in the text section for the sake of unwind info, particularly for functions marked "nounwind".

Suppose we have a function that's aarch64_pstate_sm_body or aarch64_pstate_sm_enabled.  We can statically determine whether SME is supposed to be enabled at any given point in the function.  If we encode that information into the DWARF, the unwinder/debugger can figure it out too.

For aarch64_pstate_sm_compatible, we always have the result of __arm_sme_state saved in some register.  (With one minor exception: if we're currently executing code in the function itself or __arm_sme_state, we might not have the value saved yet.  But in that case a debugger can directly inspect the PSTATE register).

So if we invent an appropriate DWARF encoding, we should be able to represent all the necessary information without requiring any code at runtime, I think.  Given that, what's the advantage of generating code to explicitly compute/spill "VG"?

kmclaughlin-arm committed:
- Prevent spill of VG for functions marked as 'nounwind'
- Added a test to sme-vg-to-stack.ll with the 'nounwind' attribute

kmclaughlin-arm commented:
Hi @efriedma-quic, thank you for taking a look at this!

> It seems weird to be generating actual instructions in the text section for the sake of unwind info, particularly for functions marked "nounwind".

In the latest commit I have made changes to ensure that functions marked with 'nounwind' will not store VG to the stack.

>
> So if we invent an appropriate DWARF encoding, we should be able to represent all the necessary information without requiring any code at runtime, I think. Given that, what's the advantage of generating code to explicitly compute/spill "VG"?

I created this patch to spill the value of VG at the beginning of the function as this is requested by the ABI, specifically `"the function's executable code must save the old value of VG to some location L before the operation that might change VG."`

However, I'm not sure why we don't instead encode this information in debug info; perhaps this is something @rsandifo-arm  might know more about?

rsandifo-arm commented:
The mechanism is intended to be more general than simply streaming VL vs non-streaming VL.  It is possible to change the non-streaming VL from one value to another, such as via a linux prctl.  In those cases there would be no way for the unwinder to recover the previous VL unless the previous VL was specifically saved.

So yes, I agree it's unusual to emit code specifically for unwinding purposes, but it seemed like the best compromise.

FWIW, there are already other situations where unwinding constrains code generation.  For example, if a big-endian function follows the SVE PCS and needs to save Z8, it must do the save using ST1D rather than STR, so that the low 64 bits (D8) are laid out as the unwinder expects.  Using ST1D requires a predicate input and has a smaller range, so it isn't the instruction that would naturally be chosen.

kmclaughlin-arm committed:
Changes to handle locally streaming functions with streaming mode changes:

 - Emit both the streaming and non-streaming value of VG in the prologue of
   functions with the aarch64_pstate_sm_body attribute.

 - Added the VGUnwindInfoPseudo node which expands to either .cfi_restore or a
   .cfi_offset depending on the value of the immediate used (0 or 1 respectively).

 - VGUnwindInfoPseudo nodes are emitted with the smstop/smstart pair around
   calls to streaming-mode functions from a locally-streaming caller. The
   .cfi_offset will save the streaming-VG value, whilst the restore sets the rule
   for VG to the same as it was at the beginning of the function (non-streaming).

 - The frame index used for the streaming VG value is saved in AArch64FunctionInfo
   so that it can be used to calculate the offset when expanding the pseudo.

 - Added the @vg_locally_streaming_fn() test to sme-vg-to-stack.ll

kmclaughlin-arm commented:
> On a related note, I'm not sure what we're generating here is really what we want for aarch64_pstate_sm_body. A aarch64_pstate_sm_body has two vector lengths: the length used for spilling the callee-saves, and the length used for local variables. If we only store the former, debuggers can't display local variables.

Thank you for raising this @efriedma-quic , it's something that was not handled correctly in my original patch. After some offline discussion about this scenario, I have made changes to the patch to ensure the correct value of VG can be recovered from locally-streaming functions. As I understand it, we need to store both the streaming and non-streaming vector length in the prologue and additionally save the streaming length before calls which require streaming-mode changes. After such calls, I believe we need to use `.cfi_restore` to set the rule for VG to the same as it was at the beginning of the function (which will be the non-streaming length).

This is based on my understanding of the specification as it is today, but I am happy to try a different implementation if this changes.

> Not sure this is actually the right choice, given the interaction with debug info. Presumably we want debuggers to be able to produce a stack trace for nounwind code.

Do you know if there is something else I could be checking to decide whether to emit VG for the nounwind case, or if it's better to remove the calls to `F.needsUnwindTableEntry()` that I added in the last commit altogether?

efriedma-quic commented:
> > On a related note, I'm not sure what we're generating here is really what we want for aarch64_pstate_sm_body. A aarch64_pstate_sm_body has two vector lengths: the length used for spilling the callee-saves, and the length used for local variables. If we only store the former, debuggers can't display local variables.
> 
> Thank you for raising this @efriedma-quic , it's something that was not handled correctly in my original patch. After some offline discussion about this scenario, I have made changes to the patch to ensure the correct value of VG can be recovered from locally-streaming functions. As I understand it, we need to store both the streaming and non-streaming vector length in the prologue and additionally save the streaming length before calls which require streaming-mode changes. After such calls, I believe we need to use `.cfi_restore` to set the rule for VG to the same as it was at the beginning of the function (which will be the non-streaming length).
> 
> This is based on my understanding of the specification as it is today, but I am happy to try a different implementation if this changes.

The sequences here seem fine.

> > Not sure this is actually the right choice, given the interaction with debug info. Presumably we want debuggers to be able to produce a stack trace for nounwind code.
> 
> Do you know if there is something else I could be checking to decide whether to emit VG for the nounwind case, or if it's better to remove the calls to `F.needsUnwindTableEntry()` that I added in the last commit altogether?

I think just remove the needsUnwindTableEntry() checks; it's more important to have consistent debug vs. non-debug codegen, vs. saving a couple instructions in the prologue.

------

On additional concern I just realized looking at the patch again: I don't think we can use cntd in non-streaming mode on targets without SVE.  (See #86743 for more discussion of SME without SVE.)

kmclaughlin-arm committed:
- Removed calls to needsUnwindTableEntry() so that VG is always
  spilled if there are streaming mode changes in the function.

- Added requiresVGSpill() to AArch64MachineFunctionInfo which returns
  true if the function has streaming mode changes and hasSVE is true.
  With this change, we will no longer spill VG at the beginning of
  functions if the target does not also have SVE.

- Removed SpilledStreamingVG flag from spillCalleeSavedRegisters.

- Rebased to include recent changes to the changeStreamingMode interface.

kmclaughlin-arm commented:
> On additional concern I just realized looking at the patch again: I don't think we can use cntd in non-streaming mode on targets without SVE. (See #86743 for more discussion of SME without SVE.)

If the target doesn't have SVE, I don't think there is a need to save VG even in functions which have streaming-mode changes. I've added a helper to AArch64MachineFunctionInfo which will ensure we only spill VG if both hasSVE and hasStreamingModeChanges are true.

efriedma-quic commented:
> If the target doesn't have SVE, I don't think there is a need to save VG even in functions which have streaming-mode changes

Sure, but "-sve" doesn't mean the target doesn't have SVE; it just means the target may or may not has SVE.  What if you compile the code with SVE disabled, and the target actually does end up having SVE at runtime?  (e.g. you compile a library in SME-but-not-SVE mode, but it gets linked into an application that uses SVE.)

kmclaughlin-arm committed:
- Emit .cfi_offset to save VG before all streaming-mode changes in a function.

- Change emitCalleeSavedGPRLocations to only emit the non-streaming location
  of VG in the prologue for locally-streaming functions.

- Move the .cfi_offset directive before the smstart/smstop.

- Added streaming-compatible tests.

kmclaughlin-arm commented:
> What if you compile the code with SVE disabled, and the target actually does end up having SVE at runtime? (e.g. you compile a library in SME-but-not-SVE mode, but it gets linked into an application that uses SVE.)

I believe checking for `hasSVE` will be sufficient for non-streaming functions which contain a call to a streaming-mode function. Even if the target does end up having SVE, there shouldn't be any relevant VG value in the function to save if it was only compiled with SME.

However, there are a couple of cases not covered here if the code is compiled without SVE, but the target has SVE:
 - Streaming caller -> non-streaming callee.
 - Streaming-compatible caller -> non-streaming callee.

It should be quite straightforward to extend this patch to cover the first of the cases above, as we can still emit `cntd` in the prologue if the function is in streaming mode. However, the second will be more involved as we can only emit the `cntd` & `.cfi_offset` if the condition guarding the `smstart`/`smstop` is true at runtime. To keep this first patch simpler, would you be happy for this to be limited to handle only the typical cases where the target has SVE?

efriedma-quic commented:
Streaming caller -> non-streaming callee works, sure.

From your description, not completely sure how you plan to handle the streaming-compatible caller -> non-streaming callee case; there's no such thing as a conditional cfi_offset, as far as I know.  I'd at least like to see the proposed assembly sequence, even if you don't implement it in this patch.

The problem case is "non-streaming SVE-enabled caller -> locally-streaming SVE-disabled callee".  The caller doesn't set VG because it doesn't know anything about streaming.  The callee can't set VG because it can't read the non-streaming vector length.  The only way I can think of to make it work, given the specified structure of the unwind data, is a runtime check for whether SVE is available.  But the ABI doesn't provide any way to check that at the moment.

I don't have a problem with splitting the patches however you think is appropriate, but we need to make sure the proposed scheme actually works before we go forward with it.

sdesmalen-arm commented:
> The problem case is "non-streaming SVE-enabled caller -> locally-streaming SVE-disabled callee". The caller doesn't set VG because it doesn't know anything about streaming. The callee can't set VG because it can't read the non-streaming vector length. The only way I can think of to make it work, given the specified structure of the unwind data, is a runtime check for whether SVE is available. But the ABI doesn't provide any way to check that at the moment.

If I understand you correctly, you're concerned about the following case:
```
__attribute__((target("+sme,+nosve")))
void c() __arm_streaming {
  ...
}

__attribute__((target("+sme,+nosve"))
void b() { c(); }

__attribute__((target("+sve")))
void a() { b(); }
```

where in `b()` it's not possible to read and save VG (since the function is compiled with +nosve and the function is not in streaming mode), so when unwinding from `c->b->a` there is no information to tell the unwinder/debugger what the value of VG is inside `a()`.

I think the critical point here is that having SME without also having SVE available in non-streaming mode is an atypical use-case for which the ABI was never really intended. SME is an Armv9-A feature which generally expects SVE to be available (in non-streaming mode), similar to how Armv8-A generally expects AdvSIMD to be available (see note in section A1.5 of the Arm Reference Manual). In LLVM we have tried to keep the two features conceptually separate so not to tie ourselves in in case this requirement ever needs to be relaxed.

If this is a use-case that ever needs supporting in the ABI, I guess this could be done by adding an extra ABI routine that safely provides the current value of VG iff SVE is available at runtime.

efriedma-quic commented:
If the ABI spec is intentionally not supporting the combination for now, that's fine, I guess?  As you note, it should be feasible to extend the ABI without breaking backward-compatibility.  Probably the compiler should report an error on the unsupported combinations, though.

kmclaughlin-arm committed:
- Add warnings to clang for unsupported cases where unwinding is not possible
  because of streaming-mode changes without SVE available.

- Fixed incorrect labels in sme-vg-to-stack.ll

kmclaughlin-arm committed:
- Fixed indentation in checkCall.

- Rebased after new SME warnings were added to SemaChecking.cpp in main.

kmclaughlin-arm commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -8001,6 +8007,22 @@ void Sema::checkCall(NamedDecl *FDecl, const FunctionProtoType *Proto,
+      if (CallerFD->hasAttr<ArmLocallyStreamingAttr>())
+        Diag(Loc, diag::warn_sme_locally_streaming_no_sve);
+
+        auto CallerStreamingTy = getArmStreamingFnType(CallerFD);
```
Fixed in latest commit

kmclaughlin-arm commented:
Gentle ping :)

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64SMEInstrInfo.td:
```diff
@@ -221,6 +224,15 @@ def : Pat<(AArch64_smstop (i32 svcr_op:$pstate), (i64 /*AArch64SME::Always*/0)),
+// the streaming value of VG around streaming-mode changes in locally-streaming
+// functions.
+def VGUnwindInfoPseudo :
+  Pseudo<(outs), (ins timm0_1:$save_restore), []>, Sched<[]>;
```
Is there a reason to make this one pseudo, instead of two?  The two operations have opposite semantics, and opcode space isn't that scarce.

kmclaughlin-arm replied:
There wasn't any particular reason for only adding only pseudo, so I've split this out into two (VGSavePseudo & VGRestorePseudo).

efriedma-quic commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3730,6 +3730,12 @@ def warn_gnu_inline_cplusplus_without_extern : Warning<
   InGroup<DiagGroup<"gnu-inline-cpp-without-extern">>;
+def warn_sme_streaming_mode_change_no_sve : Warning<
+  "function requires a streaming-mode change, unwinding is not possible without 'sve'">,
+  InGroup<AArch64SMEAttributes>;
```
This should probably be an error if it's possible to unwind: it's effectively miscompile.  Both here, and in the backend.  Add a note suggesting marking the function `noexcept`/`__attribute((nothrow))`.

If it isn't possible to unwind, you just end up with slightly inaccurate debug info, which is just annoying; probably not worth warning for that.

kmclaughlin-arm replied:
Thanks @efriedma-quic, I have changed these warnings to errors as suggested and also updated SemaChecking.cpp to only emit them if the callee is not marked with either `noexcept` or `nothrow`. I thought that we should not emit the errors if `fno-exceptions` is used as well, so I've added a check for `getLangOpts().Exceptions` in the same place.

I started adding similar asserts to LLVM to ensure that if SVE is not available, that nounwind is set on the callee. I haven't included these in the latest commit however, as many of the LLVM tests for SME do not currently pass `-mattr=+sve` and would require updating a lot of CHECK lines. I would prefer to work on this separately and post a new patch following this PR if possible, as I think including all of the test changes will make this PR more difficult to review.

kmclaughlin-arm committed:
- Split pseudo into VGSavePseudo & VGRestorePseudo.

- Check for noexcept or nothrow when emitting Clang errors.

efriedma-quic reviewed (approved):
LGTM

kmclaughlin-arm committed:
- Replace pseudos with cfi instructions in processFunctionBeforeFrameIndicesReplaced
  and removed handling from AArch64ExpandPseudoInsts.

- Removed diagnostics from Clang for unwinding without +sve.

- Removed hasSVE() check when emitting pseudos around calls in AArch64ISelLowering.

- Emit a call to __arm_get_current_vg from spillCalleeSavedRegisters if
  HasSVE is false & preserve X0 around the call if live.

- Updated LLVM tests with streaming-mode changes to also pass +sve.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -8287,6 +8289,13 @@ AArch64TargetLowering::LowerCall(CallLoweringInfo &CLI,
   SDValue InGlue;
   if (RequiresSMChange) {
+
+    if (Subtarget->hasSVE()) {
```
I'm not sure I follow what the issue is... I think we discussed before that we want debuggers to be able to unwind the stack even if a function is nounwind.

kmclaughlin-arm replied:
> when I compile some code with -fno-asynchronous-unwind-tables, it will still generate these directives. Perhaps this should have a diagnostic in Clang?

I'm happy to add such a diagnostic, but I want to make sure I have the reason for this requirement correct.
If I understand correctly, enabling asychronous unwind tables is required now that I am emitting the save and restore of VG at the point of each call in the function which changes streaming-mode.

> we want debuggers to be able to unwind the stack even if a function is nounwind.

Given this, I'm not entirely sure why we would expect to be able to unwind if `-fno-asynchronous-unwind-tables` was used or if the function has been marked as nounwind? I might have misunderstood something though!

efriedma-quic replied:
There are two different kinds of DWARF "unwind info"; one is the kind that's in a loadable section, and used for EH. The other is in a debug info section, not loaded at runtime.  If you specify `-fasynchronous-unwind-tables`, you get the former; if you specify `-g -fno-asynchronous-unwind-tables`, you get the latter.

If you request no debug info and no unwind tables, we shouldn't emit any DWARF directives.

But like I mentioned, I think we want to unconditionally emit the code to save the VG, whether or not we emit the corresponding DWARF directives.

kmclaughlin-arm replied:
If we still emit the spill of VG with `-fno-asynchronous-unwind-tables`, I don't know how we would be able to recover the correct value without the per-call CFI saves and restores. And without the correct VG value, I don't think it will be possible to recover any VG based values in the stack frame.

I'm still not sure why this wouldn't require a diagnostic, because if there is not enough information to recover VG then I don't think we can unwind correctly?

efriedma-quic replied:
If there's no DWARF unwind, nothing can unwind the stack whether or not we store the VG.

sdesmalen-arm replied:
@efriedma-quic I think the issue is that when we emit unwind info that is _not_ asynchronous, then the unwinder can't correctly unwind the stack because it would use the wrong value for VG to compute the offsets of callee-saves. So any unwind info that would be produced is broken.

efriedma-quic replied:
Oh, right, there's also that dimension.  I'm not sure I understand the interaction here, but if there's an issue, can we just force on "asynchronous" unwind info in that case?  The point of non-async unwind info isn't that it's a different unwind format; it's just an optimization to reduce the size of the unwind info.

kmclaughlin-arm replied:
I've made changes to `needsAsyncDwarfUnwindInfo` in AArch64MachineFunctionInfo.cpp to always return true if the function has streaming-mode changes. I believe this will ensure we emit the correct information when `-fno-asynchronous-unwind-tables` is specified.

efriedma-quic commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3768,6 +3768,12 @@ def err_conflicting_attributes_arm_state : Error<
   "streaming function cannot be multi-versioned">;
+def err_sme_streaming_mode_change_no_sve : Error<
+  "function requires a streaming-mode change, unwinding is not possible without 'sve'. "
+  "Consider marking this function as 'noexcept' or '__attribute__((nothrow))'">;
```
If we call new functions, they need to be part of the ABI.  If you're happy to work with your ABI people to document the new interface, I guess it's not a problem.

kmclaughlin-arm replied:
Documentation for this function was added in https://github.com/ARM-software/abi-aa/pull/263 and https://github.com/llvm/llvm-project/pull/92921 adds the routine to compiler-rt.

kmclaughlin-arm commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -1552,6 +1553,57 @@ bool AArch64ExpandPseudo::expandMI(MachineBasicBlock &MBB,
+       // This is a restore of VG after returning from the call. Emit the
+       // .cfi_restore instruction, which sets the rule for VG to the same
+       // as it was on entry to the function.
+       ++MBBI;
```
This was a mistake, the iterator shouldn't be changing here as the pseudos are emitted in the correct place around the call.

kmclaughlin-arm commented on llvm/lib/Target/AArch64/AArch64ExpandPseudoInsts.cpp:
```diff
@@ -1552,6 +1553,57 @@ bool AArch64ExpandPseudo::expandMI(MachineBasicBlock &MBB,
+               TII.get(TargetOpcode::CFI_INSTRUCTION))
+           .addCFIIndex(CFIIndex)
+           .setMIFlags(MachineInstr::FrameSetup);
+     } else {
```
Expanding the pseudos has moved to AArch64FrameLowering.cpp. There is more common code here, so I have only added one function to expand both (`emitVGSaveRestore`). I can split this into an `emitVGSave`/`emitVGRestore` if you'd still prefer they be kept separate.

kmclaughlin-arm committed:
- Added PreserveMost_From_X1 regmask to __arm_get_current_vg call

kmclaughlin-arm committed:
- Return true from needsAsyncDwarfUnwindInfo() if the function contains
  a streaming-mode change.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp:
```diff
@@ -196,12 +196,14 @@ bool AArch64FunctionInfo::needsAsyncDwarfUnwindInfo(
-                                F.getUWTableKind() == UWTableKind::Async &&
-                                !F.hasMinSize();
+    NeedsAsyncDwarfUnwindInfo =
+        (needsDwarfUnwindInfo(MF) && F.getUWTableKind() == UWTableKind::Async &&
```
If we can't emit correct async unwind info for functions with outlining/homogeneous epilogues, does that mean we also need to disable outlining/homogeneous epilogues for functions with streaming mode changes?

efriedma-quic replied:
Oh, also, this should probably be `NeedsAsyncDwarfUnwindInfo = needsDwarfUnwindInfo(MF) && ((F.getUWTableKind() == UWTableKind::Async && !F.hasMinSize()) || AFI->hasStreamingModeChanges());`.

kmclaughlin-arm replied:
This PR does include a change to `AArch64FrameLowering::homogeneousPrologEpilog` which disables homogeneous epilogues if the function has streaming-mode changes.

I hadn't considered outlining, but I can see that when considering candidates we must be able to outline all CFI instructions in the function. Am I correct in thinking that this is the reason we would need to disable outlining when there are streaming-mode changes which require async unwind info?

efriedma-quic replied:
I'm not sure exactly what the issues are with outlining at this point, but the last time async-unwind was looked at, there apparently were issues.

kmclaughlin-arm replied:
I think outlining from functions with streaming-mode changes needs more investigation. I don't think this is just a concern for async unwind; I noticed that when passing `-enable-machine-outliner` to sme-vg-to-stack.ll that some calls are outlined with only one of the smstart/smstop instructions surrounding the call. I'm not sure if this is safe yet, so for now I've disabled outlining for these functions in `isFunctionSafeToOutlineFrom`.

efriedma-quic commented on llvm/test/CodeGen/AArch64/sme-disable-gisel-fisel.ll:
```diff
@@ -214,7 +232,8 @@ declare double @za_shared_callee(double) "aarch64_inout_za"
 ; CHECK-COMMON:       // %bb.0: // %prelude
-; CHECK-COMMON-NEXT:    stp x29, x30, [sp, #-16]! // 16-byte Folded Spill
+; CHECK-COMMON-NEXT:    stp x29, x30, [sp, #-32]! // 16-byte Folded Spill
+; CHECK-COMMON-NEXT:    str x19, [sp, #16] // 8-byte Folded Spill
```
I feel like I must have asked about this at some point, but where is the x19 spill coming from?

kmclaughlin-arm replied:
The base pointer x19 is added to the list of saved registers in `determineCalleeSaves` if the function is in streaming mode or has SVE. The spill was introduced here when I enabled SVE for this test.

efriedma-quic replied:
Oh, right, it's not directly connected to this patch, it's just because you're changing the RUN line.

Please file a bug for this.

kmclaughlin-arm replied:
I've taken a closer look at this and I think the changes in `@za_new_caller_to_za_shared_callee` are correct. Because it's an `aarch64_new_za` function, it has to set up a lazy-save buffer on entry and a variable sized object is allocated on the stack for this. The function is not streaming, so x19 will only be spilled if the function has SVE.

The buffer is not used however, and https://github.com/llvm/llvm-project/pull/81648 was created to remove the lazy-save from tests like this where it is not required.

kmclaughlin-arm committed:
- Fixed misplaced brackets in needsAsyncDwarfUnwindInfo()

kmclaughlin-arm committed:
- Disable outlining from functions which require streaming-mode changes.

efriedma-quic reviewed (approved):
LGTM

kmclaughlin-arm committed:
- Rebased & updated outlining-with-streaming-mode-changes.ll test

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -3062,7 +3131,68 @@ bool AArch64FrameLowering::spillCalleeSavedRegisters(
+              .setMIFlag(MachineInstr::FrameSetup);
+        else {
+          const AArch64Subtarget &STI = MF.getSubtarget<AArch64Subtarget>();
+          for (const auto &LiveIn : MBB.liveins())
+            if (STI.getRegisterInfo()->isSuperOrSubRegisterEq(AArch64::X0,
+                                                              LiveIn.PhysReg))
+              X0Scratch = Reg1;
```
Is this missing a `break` ?

If so, you can write this using any_of, e.g.:
```
if (llvm::any_of(MBB.liveins(), [&STI](const RegisterMaskPair &LiveIn) {
      return STI.getRegisterInfo()->isSuperOrSubRegisterEq(AArch64::X0,
                                                           LiveIn.PhysReg);
    }))
  X0Scratch = Reg1;
```

sdesmalen-arm reviewed (approved):
Some very minor nits, but overall LGTM.

kmclaughlin-arm committed:
Improvements to spillCalleeSavedRegisters:
  - Removed HasSVE
  - Use any_of to check if X0 is a livein
  - Removed level of indentation from block handling VG spill
